Add required questions to downgrading subscription modal

gbova opened the pull request:
This PR updates the Downgrading? modal that pops up when a user downgrades their subscription. Growth needs better data on why people downgrade.

Specs:
https://docs.google.com/document/d/1wdN02xuGMXuLPoQ3QJ16zm_WNNA2-u8yDl5kFxVGzCM/edit
https://docs.google.com/document/d/13U-VQSF5ODNd-3QbDgkeDekBTwroG6Fl6BEnwfsuSJg/edit

Review commit-by-commit

gbova committed:
Add new questions to downgrade modal

gbova committed:
Send downgrade information to confirm billing page

gbova committed:
Send downgrade email after subscription is changed

gbova committed:
Make downgrade modal questions required

orangejenny commented:
screenshot?

orangejenny commented on corehq/apps/accounting/static/accounting/js/confirm_plan.js:
```diff
@@ -17,9 +17,35 @@ hqDefine('accounting/js/confirm_plan', [
+
+
+        self.downgradeReasonList = ko.observableArray([
+            "My project ended",
```
`gettext` for these

orangejenny commented on corehq/apps/accounting/static/accounting/js/confirm_plan.js:
```diff
@@ -17,9 +17,35 @@ hqDefine('accounting/js/confirm_plan', [
+            return self.downgradeReason() === "My project ended";
+        });
+        self.newToolNeeded = ko.computed(function () {
+            return self.downgradeReason() === "We are switching to a different mobile data collection tool";
```
I'd make the reasons constants; they're fairly long strings so these computed observables are subject to getting out of sync with the values above if someone, say, changes the wording.

orangejenny commented on corehq/apps/accounting/static/accounting/js/confirm_plan.js:
```diff
@@ -17,9 +17,35 @@ hqDefine('accounting/js/confirm_plan', [
 
+
+
+        self.downgradeReasonList = ko.observableArray([
```
I think these lists of reasons can be regular arrays - they'd need to be observables if you were adding/removing objects and needed to watch those changes.

gbova commented:
Thanks for the reminder @orangejenny !

<img width="1439" alt="screen shot 2018-09-14 at 5 03 01 pm" src="https://user-images.githubusercontent.com/15271571/45575064-2f8d3080-b840-11e8-9947-f0b1c5fdf793.png">
<img width="1438" alt="screen shot 2018-09-14 at 5 03 16 pm" src="https://user-images.githubusercontent.com/15271571/45575065-2f8d3080-b840-11e8-951d-554ebd4e8c45.png">
<img width="1437" alt="screen shot 2018-09-14 at 5 03 25 pm" src="https://user-images.githubusercontent.com/15271571/45575066-2f8d3080-b840-11e8-9d53-47cc7b1f70f5.png">

orangejenny commented on corehq/apps/accounting/static/accounting/js/confirm_plan.js:
```diff
@@ -57,25 +57,30 @@ hqDefine('accounting/js/confirm_plan', [
-            });
+            if (self.form) {
+                var downgradeReason = "";
+                var selectedDowngradeOptions = document.getElementById("select-downgrade-reason").selectedOptions;
```
Style nitpick: it'd be more concise to do something like $("#select-downgrade-reason").val().join(", ")

orangejenny commented on corehq/apps/accounting/static/accounting/js/confirm_plan.js:
```diff
@@ -45,6 +48,19 @@ hqDefine('accounting/js/confirm_plan', [
             return self.newToolReason() === "Other";
         });
+        self.requiredQuestionsAnswered = ko.computed(function () {
+            if (self.downgradeReason() == null) {
```
I'd lean towards making this `if (!self.downgradeReason())` to protect against code changes that make downgradeReason `undefined`.

orangejenny commented on corehq/apps/accounting/static/accounting/js/confirm_plan.js:
```diff
@@ -17,9 +17,51 @@ hqDefine('accounting/js/confirm_plan', [
+        self.newTool = ko.observable("");
+        self.newToolReason = ko.observable("");
+        self.otherNewToolReason = ko.observable("");
+        self.requiredQuestionsAnswered = ko.observable(false);
```
fyi there's a knockout validation plugin in our codebase - you'd add `@use_ko_validation` to the python view, use `.extend` on the required observables as described in [the docs](https://github.com/Knockout-Contrib/Knockout-Validation#getting-started), and use our custom [koValidationStateFeedback`(https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/js/validators.ko.js) binding to display messages. Totally optional; I haven't used it enough to be sure whether it would be overkill for the amount of validation here or would be an improvement.

gbova replied:
This does seem really cool, and I'm playing with it now! But, I don't think that it's necessary for this change right now.

gbova committed:
Use constants for downgrade answers

gbova committed:
Lint

gbova committed:
Clean up loops to get multiple downgrading answers

gbova committed:
Check for null/empty with !thing()

I couldn't figure out a simple way of converting line 65: var newToolAnswered = self.newTool() !== "";

gbova committed:
Do not send downgrading email for superusers

gbova committed:
Use gettext()

gbova committed:
Use regular arrays for downgrading answers
